src/Microsoft.ML.Data/DataLoadSave/Binary/Codecs.cs

```diff
@@ -657,124 +655,100 @@ public Reader(DateTimeCodec codec, Stream stream, int items)
                 public override void MoveNext()
                 {
                     Contracts.Assert(_remaining > 0, "already consumed all values");
-                    var value = Reader.ReadInt64();
-                    Contracts.CheckDecode(value == DvInt8.RawNA || (ulong)value <= DvDateTime.MaxTicks);
-                    _value = new DvDateTime(value);
+                    _value = new DateTime(Reader.ReadInt64());
                     _remaining--;
                 }
 
-                public override void Get(ref DvDateTime value)
+                public override void Get(ref DateTime value)
                 {
                     value = _value;
                 }
             }
         }
 
-        private sealed class DateTimeZoneCodec : SimpleCodec<DvDateTimeZone>
+        private sealed class DateTimeOffsetCodec : SimpleCodec<DateTimeOffset>
         {
-            private readonly MadeObjectPool<short[]> _shortBufferPool;
             private readonly MadeObjectPool<long[]> _longBufferPool;
 
-            public DateTimeZoneCodec(CodecFactory factory)
-                : base(factory, DateTimeZoneType.Instance)
+            public DateTimeOffsetCodec(CodecFactory factory)
+                : base(factory, DateTimeOffsetType.Instance)
             {
-                _shortBufferPool = new MadeObjectPool<short[]>(() => null);
                 _longBufferPool = new MadeObjectPool<long[]>(() => null);
             }
 
-            public override IValueWriter<DvDateTimeZone> OpenWriter(Stream stream)
+            public override IValueWriter<DateTimeOffset> OpenWriter(Stream stream)
             {
                 return new Writer(this, stream);
             }
 
-            public override IValueReader<DvDateTimeZone> OpenReader(Stream stream, int items)
+            public override IValueReader<DateTimeOffset> OpenReader(Stream stream, int items)
             {
                 return new Reader(this, stream, items);
             }
 
-            private sealed class Writer : ValueWriterBase<DvDateTimeZone>
+            private sealed class Writer : ValueWriterBase<DateTimeOffset>
             {
-                private List<short> _offsets;
```

Comment:
I'm not sure on how concerned we need to be with sizes, but an offset can only be in the range of `-14` to `14` hours. So using a `long` seems like a bit of an overkill.

It also appears that previously the `_offsets` was the number of minutes in the offset.  Now we are using the number of ticks in the offset.  Do we need to be concerned with this difference?

Reply:
.NET standard type has offsets in ticks. We could convert offset in minutes when writing to disk and convert it back to ticks when reading. What do you think?

---
In reply to: [211667357](https://github.com/dotnet/machinelearning/pull/693#discussion_r211667357) [](ancestors = 211667357)

Reply:
`DateTimeOffset` exposes its offset as a `TimeSpan`, but internally it uses `short` and in minutes.

[https://github.com/dotnet/coreclr/blob/9499b08eefd895158c3f3c7834e185a73619128d/src/System.Private.CoreLib/shared/System/DateTimeOffset.cs#L51-L53](https://github.com/dotnet/coreclr/blob/9499b08eefd895158c3f3c7834e185a73619128d/src/System.Private.CoreLib/shared/System/DateTimeOffset.cs#L51-L53)

[https://github.com/dotnet/coreclr/blob/9499b08eefd895158c3f3c7834e185a73619128d/src/System.Private.CoreLib/shared/System/DateTimeOffset.cs#L286-L292](https://github.com/dotnet/coreclr/blob/9499b08eefd895158c3f3c7834e185a73619128d/src/System.Private.CoreLib/shared/System/DateTimeOffset.cs#L286-L292)

From everything I can find online (ISO8601, [RFC3339](https://tools.ietf.org/html/rfc3339), [SQL Server doc](https://docs.microsoft.com/en-us/sql/t-sql/data-types/datetimeoffset-transact-sql?view=sql-server-2017#time-zone-offset)), the offset supports the range `-14` to `14` hours, and only supports minute precision.

---
In reply to: [212488658](https://github.com/dotnet/machinelearning/pull/693#discussion_r212488658) [](ancestors = 212488658,211667357)

Reply:
Thanks a lot! this is helpful. I'm now writing the ticks in minutes as short values because they are already validated when TimeSpan is created and when reading I convert the minutes back into Ticks when creating the TimeSpan object.

---
In reply to: [212670585](https://github.com/dotnet/machinelearning/pull/693#discussion_r212670585) [](ancestors = 212670585,212488658,211667357)